File: 学校/信息技术/出题/4/std1.cpp
#include <cstdio>

const int N = 1e3 + 10;

int a[N];

int main() {
  int n, q;
  scanf("%d%d", &n, &q);
  for (int i = 1; i <= n; ++i)
    scanf("%d", &a[i]);
  int last_ans = 0;
  while (q--) {
    int ll, rr, ans = 0, l, r;
    scanf("%d%d", &ll, &rr);
    ll = (ll ^ last_ans - 1 + n) % n + 1;
    rr = (rr ^ last_ans - 1 + n) % n + 1;
    if (ll > rr) ll ^= rr ^= ll ^= rr;
    l = ll, r = rr;
    for (int i = l; i <= r; ++i)
      for (int j = i + 1; j <= r; ++j)
        if (a[i] > a[j]) ans += 1;
    printf("%d\n", last_ans = ans);
  }
  return 0;
}

File: 学校/信息技术/出题/4/std1.py
from bisect import bisect_left

n, q = map(int, input().split())
a = list(map(int, input().split()))
b = list(set(a)); b.sort()
for i in range(n):
  a[i] = bisect_left(b, a[i]) + 1
qwq = [0] * (n + 1)
def lowbit(i): return i & (-i)
def add(p):
  while (p <= n):
    qwq[p] += 1
    p += lowbit(p)
def sum(p):
  res = 0
  while (p):
    res += qwq[p]
    p -= lowbit(p)
  return res
last_ans = 0
while (q):
  q -= 1
  ans = 0
  l, r = map(int, input().split())
  ll = (l ^ last_ans - 1) % n + 1
  rr = (r ^ last_ans - 1) % n + 1
  if (ll > rr): ll, rr = rr, ll
  l, r = ll, rr
  l -= 1; r -= 1
  for i in range(0, n + 1): qwq[i] = 0
  for i in range(l, r + 1):
    ans += (i - l) - sum(a[i]); add(a[i])
  last_ans = ans
  print(ans)